File: solutions/0241. Different Ways to Add Parentheses/0241.cpp
class Solution {
 public:
  vector<int> diffWaysToCompute(string expression) {
    unordered_map<string, vector<int>> memo;
    return ways(expression, memo);
  }

 private:
  vector<int> ways(const string& s, unordered_map<string, vector<int>>& memo) {
    if (memo.count(s))
      return memo[s];

    vector<int> ans;

    for (int i = 0; i < s.length(); ++i)
      if (ispunct(s[i]))
        for (const int a : ways(s.substr(0, i), memo))
          for (const int b : ways(s.substr(i + 1), memo))
            if (s[i] == '+')
              ans.push_back(a + b);
            else if (s[i] == '-')
              ans.push_back(a - b);
            else
              ans.push_back(a * b);

    return memo[s] = (ans.empty() ? vector<int>{stoi(s)} : ans);
  }
};

File: solutions/0241. Different Ways to Add Parentheses/0241.py
class Solution:
  @lru_cache(None)
  def diffWaysToCompute(self, expression: str) -> List[int]:
    ans = []

    for i, c in enumerate(expression):
      if c in '+-*':
        left = self.diffWaysToCompute(expression[:i])
        right = self.diffWaysToCompute(expression[i + 1:])
        for a in left:
          for b in right:
            ans.append(eval(str(a) + c + str(b)))

    return ans or [int(expression)]
